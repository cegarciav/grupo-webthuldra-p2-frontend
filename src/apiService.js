import axios from 'axios';

const apiOrigin = process.env.REACT_APP_API_ORIGIN;

const apiRequester = axios.create({
  baseURL: apiOrigin,
});

const errorGenerator = (e) => {
  let errorResponse = null;
  if (e.response) {
    errorResponse = {
      errors: e.response.data.errors,
      type: 'response',
      statusCode: e.response.status,
    };
  } else {
    errorResponse = {
      type: 'unexpected',
      message: 'Algo salió mal. Por favor, revisa tu conexión o intenta ingresar más tarde',
    };
  }
  return errorResponse;
};

async function apiGet(path, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.get(
      url, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

async function apiPost(path, body, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.post(
      url,
      body, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

async function apiPatch(path, body, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.patch(
      url,
      body, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

async function apiDelete(path, body, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.delete(
      url,
      body, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

export {
  apiGet,
  apiPost,
  apiPatch,
  apiDelete,
};
