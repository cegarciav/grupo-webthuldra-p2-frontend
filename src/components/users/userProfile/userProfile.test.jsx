import React from 'react';
import PropTypes from 'prop-types';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { cleanup } from '@testing-library/react';
import MainRouter from '../../../routers/index';
import { apiGet } from '../../../apiService';

function TestRouter({ path }) {
  return (
    <MemoryRouter initialEntries={[path]}>
      <MainRouter />
    </MemoryRouter>
  );
}

TestRouter.propTypes = {
  path: PropTypes.string.isRequired,
};

jest.mock('../../../apiService');

const userId = '96ab7b63-915e-442b-b83f-e0962bd2c8d8';

const userResponse = {
  firstName: 'User Name 1',
  lastName: 'User Lastname 1',
  email: 'user@email.com',
  id: userId,
};

const commentsResponse = [
  {
    id: '26c08e2d-3caf-458b-9767-8477683668c8',
    text: 'Possitive comment',
    grade: 4,
    reviewerId: userId,
    storeId: '3e7ca498-8646-4493-bce6-b8cad8c5908e',
    store: {
      name: 'Name 1',
      address: 'Address 1',
    },
    reviewer: {
      firstName: 'User Name 1',
      lastName: 'User Lastname 1',
      email: 'user@email.com',
    },
  },
  {
    id: '26c08e2d-3caf-458b-9767-8477683668c9',
    text: 'Negative comment',
    grade: 1,
    reviewerId: userId,
    storeId: '3e7ca498-8646-4493-bce6-b8cad8c5908e',
    store: {
      name: 'Name 1',
      address: 'Address 1',
    },
    reviewer: {
      firstName: 'User Name 1',
      lastName: 'User Lastname 1',
      email: 'user@email.com',
    },
  },
];

const dealsResponse = [
  {
    id: '2927c6bb-d0c4-4583-b75f-e21f1a30b497',
    status: 'abierto',
    customerId: '96ab7b63-915e-442b-b83f-e0962bd2c8d8',
    storeId: '6f4e5985-2ff8-4118-bbaf-e6004d757c13',
    customer: {
      firstName: 'User Name 1',
      lastName: 'User Lastname 1',
      email: 'user@email.com',
      id: '96ab7b63-915e-442b-b83f-e0962bd2c8d8',
    },
    store: {
      name: 'Tienda 2',
      id: '6f4e5985-2ff8-4118-bbaf-e6004d757c13',
    },
    products: [
      {
        name: 'Product 2',
        price: 500,
        purchase: {
          dealId: '2927c6bb-d0c4-4583-b75f-e21f1a30b497',
          productId: 'a9aa09e7-83f8-4e95-b551-273cedf0afe3',
          amount: 4,
        },
      },
    ],
  },
];

const storesResponse = [
  {
    id: '6f4e5985-2ff8-4118-bbaf-e6004d757c13',
    address: 'Address 1',
    name: 'Store name 1',
    description: 'Description 1',
    ownerId: userId,
    picture: null,
  },
  {
    id: '9440a98a-2973-484b-b8bd-78beb6b24eef',
    address: 'Address 3',
    name: 'Store name 3',
    description: 'Description 3',
    ownerId: userId,
    picture: null,
  },
];

describe('Users profile', () => {
  describe('When user is not logged-in', () => {
    it('renders the profile of a given user', async () => {
      let storesRenderer;
      apiGet.mockImplementation((url) => {
        if (url === `/users/${userId}`) {
          return {
            data: userResponse,
            statusCode: 200,
          };
        }
        if (url === '/stores') {
          return {
            data: storesResponse,
            statusCode: 200,
          };
        }
        return {};
      });
      await renderer.act(async () => {
        storesRenderer = await renderer.create(<TestRouter path={`/users/${userId}`} />);
      });
      const tree = storesRenderer.toJSON();
      expect(tree).toMatchSnapshot();
    });
    afterEach(() => {
      cleanup();
      jest.resetAllMocks();
    });
  });

  describe('When user is logged-in and check their own profile', () => {
    it('renders the profile of the logged-in user', async () => {
      let storesRenderer;
      apiGet.mockImplementation((url) => {
        if (url === '/users/me') {
          return {
            data: userResponse,
            statusCode: 200,
          };
        }
        if (url === '/stores') {
          return {
            data: storesResponse,
            statusCode: 200,
          };
        }
        if (url === '/users/me/comments') {
          return {
            data: commentsResponse,
            statusCode: 200,
          };
        }
        if (url === '/users/me/deals') {
          return {
            data: dealsResponse,
            statusCode: 200,
          };
        }
        return {};
      });
      await renderer.act(async () => {
        storesRenderer = await renderer.create(<TestRouter path="/users/my-profile" />);
      });
      const tree = storesRenderer.toJSON();
      expect(tree).toMatchSnapshot();
    });
    afterEach(() => {
      cleanup();
      jest.resetAllMocks();
    });
  });
});
