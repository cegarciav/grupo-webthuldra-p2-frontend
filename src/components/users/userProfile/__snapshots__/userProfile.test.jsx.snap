// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users profile When user is logged-in and check their own profile renders the profile of the logged-in user 1`] = `
<main>
  <section
    className="user-header-content"
  >
    <figure
      className="profile-pic"
    >
      <img
        alt="User Name 1 User Lastname 1"
        src="/default_profile_picture.png"
      />
      <figcaption>
        <h1>
          User Name 1 User Lastname 1
        </h1>
        <h4>
          Contacto: user@email.com
        </h4>
      </figcaption>
    </figure>
    <section
      className="user-profile-buttons"
    >
      <a
        href="/users/me/edit"
        role="button"
      >
        Editar Perfil
      </a>
      <a
        href="/users/me/delete"
        role="button"
      >
        Eliminar Perfil
      </a>
    </section>
  </section>
  <section
    className="user-main-profile"
  >
    <aside
      className="user-info"
    >
      <section
        className="user-info-section"
      >
        <main
          className="deals-container"
        >
          <h4>
            Deals
          </h4>
          <ul
            className="deals-list"
          >
            <li>
              <article
                className="comment-summary"
              >
                <section
                  className="user-info"
                >
                  <a
                    href="/stores/6f4e5985-2ff8-4118-bbaf-e6004d757c13"
                    onClick={[Function]}
                  >
                    <figure>
                      <img
                        alt="Tienda 2"
                        src="/default_store_picture.png"
                      />
                    </figure>
                    <p>
                      abierto
                    </p>
                  </a>
                </section>
                <section
                  className="comment-text"
                >
                  <a
                    href="/stores/6f4e5985-2ff8-4118-bbaf-e6004d757c13/deals/2927c6bb-d0c4-4583-b75f-e21f1a30b497"
                    onClick={[Function]}
                  >
                    Tienda 2
                  </a>
                </section>
              </article>
            </li>
          </ul>
        </main>
      </section>
      <section
        className="user-info-section"
      >
        <main
          className="comments-container"
        >
          <h4>
            Comentarios
          </h4>
          <ul
            className="comments-list"
          >
            <li>
              <article
                className="comment-summary"
              >
                <section
                  className="user-info"
                >
                  <a
                    href="/users/96ab7b63-915e-442b-b83f-e0962bd2c8d8"
                    onClick={[Function]}
                  >
                    <figure>
                      <img
                        alt="User Name 1 User Name 1"
                        src="/default_profile_picture.png"
                      />
                    </figure>
                    <p>
                      4/5
                    </p>
                  </a>
                </section>
                <section
                  className="comment-text"
                >
                  Possitive comment
                </section>
              </article>
            </li>
            <li>
              <article
                className="comment-summary"
              >
                <section
                  className="user-info"
                >
                  <a
                    href="/users/96ab7b63-915e-442b-b83f-e0962bd2c8d8"
                    onClick={[Function]}
                  >
                    <figure>
                      <img
                        alt="User Name 1 User Name 1"
                        src="/default_profile_picture.png"
                      />
                    </figure>
                    <p>
                      1/5
                    </p>
                  </a>
                </section>
                <section
                  className="comment-text"
                >
                  Negative comment
                </section>
              </article>
            </li>
          </ul>
        </main>
      </section>
    </aside>
    <section
      className="stores-in-user"
    >
      <a
        className="new-store-button"
        href="/stores/new"
        role="button"
      >
        Crear tienda
      </a>
      <main>
        <h1>
          Lista de tiendas
        </h1>
        <section
          className="searchbar-container"
        >
          <input
            className="main-searchbar"
            onChange={[Function]}
            placeholder="Buscar tiendas"
            type="text"
            value=""
          />
        </section>
        <ul
          className="stores-list"
        >
          <li>
            <a
              className="store-summary"
              href="/stores/6f4e5985-2ff8-4118-bbaf-e6004d757c13"
              onClick={[Function]}
            >
              <article>
                <h1>
                  Store name 1
                </h1>
                <section>
                  <img
                    alt="Description 1"
                    src="/default_store_picture.png"
                  />
                </section>
                <p>
                  Direcci贸n: Address 1
                </p>
              </article>
            </a>
          </li>
          <li>
            <a
              className="store-summary"
              href="/stores/9440a98a-2973-484b-b8bd-78beb6b24eef"
              onClick={[Function]}
            >
              <article>
                <h1>
                  Store name 3
                </h1>
                <section>
                  <img
                    alt="Description 3"
                    src="/default_store_picture.png"
                  />
                </section>
                <p>
                  Direcci贸n: Address 3
                </p>
              </article>
            </a>
          </li>
        </ul>
      </main>
    </section>
  </section>
</main>
`;

exports[`Users profile When user is not logged-in renders the profile of a given user 1`] = `
<main>
  <section
    className="user-header-content"
  >
    <figure
      className="profile-pic"
    >
      <img
        alt="User Name 1 User Lastname 1"
        src="/default_profile_picture.png"
      />
      <figcaption>
        <h1>
          User Name 1 User Lastname 1
        </h1>
        <h4>
          Contacto: user@email.com
        </h4>
      </figcaption>
    </figure>
    <section
      className="user-profile-buttons"
    >
      <a
        href="/users/me/edit"
        role="button"
      >
        Editar Perfil
      </a>
      <a
        href="/users/me/delete"
        role="button"
      >
        Eliminar Perfil
      </a>
    </section>
  </section>
  <section
    className="user-main-profile"
  >
    <section
      className="stores-in-user"
    >
      <a
        className="new-store-button"
        href="/stores/new"
        role="button"
      >
        Crear tienda
      </a>
      <main>
        <h1>
          Lista de tiendas
        </h1>
        <section
          className="searchbar-container"
        >
          <input
            className="main-searchbar"
            onChange={[Function]}
            placeholder="Buscar tiendas"
            type="text"
            value=""
          />
        </section>
        <ul
          className="stores-list"
        >
          <li>
            <a
              className="store-summary"
              href="/stores/6f4e5985-2ff8-4118-bbaf-e6004d757c13"
              onClick={[Function]}
            >
              <article>
                <h1>
                  Store name 1
                </h1>
                <section>
                  <img
                    alt="Description 1"
                    src="/default_store_picture.png"
                  />
                </section>
                <p>
                  Direcci贸n: Address 1
                </p>
              </article>
            </a>
          </li>
          <li>
            <a
              className="store-summary"
              href="/stores/9440a98a-2973-484b-b8bd-78beb6b24eef"
              onClick={[Function]}
            >
              <article>
                <h1>
                  Store name 3
                </h1>
                <section>
                  <img
                    alt="Description 3"
                    src="/default_store_picture.png"
                  />
                </section>
                <p>
                  Direcci贸n: Address 3
                </p>
              </article>
            </a>
          </li>
        </ul>
      </main>
    </section>
  </section>
</main>
`;
